lib/collection/src/shards/replica_set/shard_transfer.rs

```diff
@@ -207,4 +201,52 @@ impl ShardReplicaSet {
             )))
         }
     }
+
+    /// Send all queue proxy updates to remote and transform into forward proxy
+    ///
+    /// When a queue or forward proxy shard needs to be unproxified into a local shard again we
+    /// typically don't have room to handle errors. A queue proxy shard may error if it fails to
+    /// send updates to the remote shard, while a forward proxy does not fail at all when
+    /// transforming.
+    ///
+    /// This method allows to transfer queued updates before the shard is unproxified. This allows
+    /// for proper error handling at the time this method is called. Because the shard is
+    /// transformed into a forward proxy after this operation it will not error again when the
+    /// shard is eventually unproxified again.
+    ///
+    /// If the local shard is a queue proxy:
+    /// - Transfers all missed updates to remote
+    /// - Transforms queue proxy into forward proxy
+    ///
+    /// Does nothing if the local shard is not a queue proxy.
+    ///
+    /// # Errors
+    ///
+    /// Returns an error if transferring all updates to the remote failed.
+    pub async fn queue_proxy_into_forward_proxy(&self) -> CollectionResult<()> {
+        // First pass: transfer all missed updates with shared read lock
+        {
+            let local_read = self.local.read().await;
+            if let Some(Shard::QueueProxy(proxy)) = &*local_read {
+                proxy.transfer_all_missed_updates().await?;
+            } else {
+                return Ok(());
+            }
+        }
+
+        // Second pass: transfer new updates, safely finalize and transform
+        let mut local_write = self.local.write().await;
+        match local_write.take() {
+            Some(Shard::QueueProxy(proxy)) => {
+                let (local_shard, remote_shard) = proxy.finalize().await.map_err(|(err, _)| err)?;
+
```

Comment:
Let's handle error by returning proxy back into `local_write`, otherwise it might be lost in case of transfer error

Reply:
Resolved in b0c876465d927dda223c78bad84ff2470634b3d4.